# Vector Operations

I will be using the `cereal.csv` file, which was used in Chapter 3, as my data source. 


In [2]:
# import libraries
import pandas as pd
import numpy as np # this is used for numerical computing
import numpy.typing as npt # this is used for type annotation

In [3]:
# load data into dataframe
df = pd.read_csv('../datasets/cereal.csv')


Before I begin, I'd like to explain what vectors are.

Vectors are ordered list of numbers used to represent quantities or data mathematically. 

Vectors are the fundamental way machine learning represents and manipulates data. They are 1-dimensional structures.

Vectors are usually represented using NumPy arrays. (_Note that this does not mean all arrays are vectors_)

In this project, I will be representing various columns as vectors, and perform vector operations on them.

In [4]:
# represent the 'sugars' and 'calories' columns as  vectors
sugar_vector = np.array(df['sugars'])
calories_vector = np.array(df['calories'])

# check that they are both arrays
print(f"Sugar Vector is a {type(sugar_vector)}")
print(f"Calories Vector is a {type(calories_vector)}")

Sugar Vector is a <class 'numpy.ndarray'>
Calories Vector is a <class 'numpy.ndarray'>


The first question I'll be asking is: "Do cereals that have more sugar also tend to have more calories?"

To answer this question, I will be using Cosine Similarity

In [5]:
# create function to calculate cosine similarity
def cosine_similarity(A: npt.NDArray[np.float64], B: npt.NDArray[np.float64]) -> float:
    # compute dot product
    dot_prod = np.dot(A, B)

    # compute the norm of the vectors
    norm_A = np.linalg.norm(A)
    norm_B = np.linalg.norm(B)

    # find the product of the norms
    vector_norm = norm_A * norm_B

    # compute cosine similarity
    cos_sim = np.round(dot_prod / vector_norm, 3)

    return(cos_sim)



In [6]:
# print cosine similarity
cos_sim = cosine_similarity(sugar_vector, calories_vector)

print(f"The Cosine Similarity of the Sugar column and Calories column is {cos_sim}")

The Cosine Similarity of the Sugar column and Calories column is 0.883


_What does Cosine Similarity even mean?_

Cosine Similarity measures how similar two data items are by calculating the cosine of the angle between them. It focuses on the direction rather than magnitude, with a score of 1 meaning identical, 0 unrelated, and -1 opposite.


The Cosine Similarity of `0.883` means that there is a strong positive similarity between the two vectors.

This may suggest that cereals with higher sugar content generally have higher calorie values.


**NOTE:** 

Cosine Similarity does not measure correlation. Cosine similarity measures geometric similarity, while correlation measures statistical relationship.



**We have answered the question!**

Cereals that have more sugar tend to have more calories.

In [7]:
one = np.array(df.iloc[5][['calories', 'protein', 'fat', 'sodium', 'carbo', 'sugars', 'potass', 'vitamins']].astype(float))
two = np.array(df.iloc[11][['calories', 'protein', 'fat', 'sodium', 'carbo', 'sugars', 'potass', 'vitamins']].astype(float))
nutrients = ['calories', 'protein', 'fat', 'sodium', 'fiber' ,'carbo', 'sugars', 'potass', 'vitamins']
measurement = ['calories', 'grams', 'grams', 'milligrams', 'grams' ,'grams', 'grams', 'milligrams', 'milligrams']
diff = one - two
cereal_list = df['name'].to_list()
cereals = list(zip(cereal_list, range(1, len(cereal_list) + 1)))
for cereal in cereals:
    print(cereal[0])

for n, d, m in zip(nutrients, diff, measurement):
    for cereal in cereals:
        if d > 0:
            print(f"{cereal[0]} has {abs(d)} {m} more {n} than ordinary Cheerios")
        elif d < 0:
            print(f"Apple Cinnamon Cheerios has {abs(d)} {m} less {n} than ordinary Cheerios")
        else:
            print(f"Apple Cinnamon Cheerios has the same amount of {n} as ordinary Cheerios")
        

100% Bran
100% Natural Bran
All-Bran
All-Bran with Extra Fiber
Almond Delight
Apple Cinnamon Cheerios
Apple Jacks
Basic 4
Bran Chex
Bran Flakes
Cap'n'Crunch
Cheerios
Cinnamon Toast Crunch
Clusters
Cocoa Puffs
Corn Chex
Corn Flakes
Corn Pops
Count Chocula
Cracklin' Oat Bran
Cream of Wheat (Quick)
Crispix
Crispy Wheat & Raisins
Double Chex
Froot Loops
Frosted Flakes
Frosted Mini-Wheats
Fruit & Fibre Dates; Walnuts; and Oats
Fruitful Bran
Fruity Pebbles
Golden Crisp
Golden Grahams
Grape Nuts Flakes
Grape-Nuts
Great Grains Pecan
Honey Graham Ohs
Honey Nut Cheerios
Honey-comb
Just Right Crunchy  Nuggets
Just Right Fruit & Nut
Kix
Life
Lucky Charms
Maypo
Muesli Raisins; Dates; & Almonds
Muesli Raisins; Peaches; & Pecans
Mueslix Crispy Blend
Multi-Grain Cheerios
Nut&Honey Crunch
Nutri-Grain Almond-Raisin
Nutri-grain Wheat
Oatmeal Raisin Crisp
Post Nat. Raisin Bran
Product 19
Puffed Rice
Puffed Wheat
Quaker Oat Squares
Quaker Oatmeal
Raisin Bran
Raisin Nut Bran
Raisin Squares
Rice Chex
Rice Kr

In [ ]:
def vector_difference(cereal_1: str, cereal_2: str) -> dict[str, str]:
    """Calculates the difference between two vectors. """

    # Filter data for selected cereals, keep only numeric columns, then convert to NumPy float array
    cereal_1 = np.array(df[df["name"] == cereal_1]
                        .select_dtypes(include='number')).astype(float)
    cereal_2 = np.array(df[df["name"] == cereal_2]
                        .select_dtypes(include='number')).astype(float)

    print(cereal_1)

vector_difference(cereal_1="Cheerios", cereal_2="Froot Loops")

[[110.         6.         2.       290.         2.        17.
    1.       105.        25.         1.         1.         1.25
   50.764999]]


In [9]:
# convert data frames to array
cereal_1 = np.array(df.iloc[5].drop(['name', 'mfr', 'type'])).astype(float)
cereal_2 = np.array(df.iloc[11].drop(['name', 'mfr', 'type'])).astype(float)

print(cereal_1 - cereal_2)

df.columns

[   0.         -4.          0.       -110.         -0.5        -6.5
    9.        -35.          0.          0.          0.         -0.5
  -21.255458]


Index(['name', 'mfr', 'type', 'calories', 'protein', 'fat', 'sodium', 'fiber',
       'carbo', 'sugars', 'potass', 'vitamins', 'shelf', 'weight', 'cups',
       'rating'],
      dtype='object')